Remove skip from inherited profiles from showing up in reporting (breaking change)

miah opened the pull request:
I've made a few changes to allow inherited profiles to be skipped when they are loaded on unsupported platforms. We will now also output a WARN to STDERR detailing which profile was skipped:

```
[2018-08-28T14:36:37-07:00] WARN: Skipping profile: 'windows-only' on unsupported platform: 'mac_os_x'.
```

Please let me know how this implementation looks and if I missed anything.

ref: https://github.com/inspec/inspec/issues/3158

miah committed:
Skip loading dependency profiles if they are unsupported on the current
platform.

Skip loading dependencies if they are unsupported on the current
platform.

Wrap our log and next in a conditional checking if the platform is
supported.

Change a `if !` into a `unless`

Check if the backend is a Train Mock Connection and if so say that the
profile does support the platform.

While iterating through tests being loaded skip when the platform is
unsupported.

We now log a WARN when a profile is skipped due to unsupported platform,
so lets check that.

Modified existing test to log that there are 0 skipped tests, instead of
2.

Add functional test that loads profile-support-skip with a json reporter
to check that our controls are not loaded and that stderr contains our
warning.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

jquick commented on lib/inspec/profile.rb:
```diff
@@ -162,10 +168,11 @@ def params
         locked_dependencies.each(&:collect_tests)
 
         tests.each do |path, content|
+          next unless supports_platform?
```
I think you could put this before the tests array like:
```
return unless supports_platform?

tests.each do |path, content|
...
```
This will keep it from being called for every single test

miah replied:
Oh right. Thank you =)

jquick reviewed (commented):
Looking really good @miah !

miah committed:
Rather than iterating through each test return before recursion if the platform is
unsupported.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe commented on test/unit/profiles/profile_test.rb:
```diff
@@ -99,6 +99,15 @@
   end
 
+  describe 'skips loading on unsupported platform' do
+    let(:profile_id) { 'windows-only' }
```
If you only use it once, my preference would be to just inline this string, and not use `let`.  Not a blocker.

clintoncwolfe commented on test/unit/profiles/profile_test.rb:
```diff
@@ -99,6 +99,15 @@
   end
 
+  describe 'skips loading on unsupported platform' do
+    let(:profile_id) { 'windows-only' }
```
I think we should add a test reflecting @frezbo's concern - a run on a linux platform should not skip a profile that supports windows and linux.  That may already be covered in another test - if so, link it and I'll be quiet :)

miah replied:
I'll add a example profile to our mocks that supports bsd, linux, and windows.

clintoncwolfe commented on test/functional/inspec_exec_test.rb:
```diff
@@ -301,7 +301,8 @@
     it 'skips the controls from that profile' do
       out = inspec("exec #{File.join(profile_path, 'profile-support-skip')} --no-create-lockfile")
-      out.stdout.force_encoding(Encoding::UTF_8).must_include "Profile Summary: 0 successful controls, 0 control failures, \e[38;5;247m2 controls skipped\e[0m\n"
+      out.stdout.force_encoding(Encoding::UTF_8).must_include "WARN: Skipping profile: 'windows-only' on unsupported platform"
```
Personally, I prefer error messages that are telling me I did something wrong, to give hints as to how to do it right; here, you could add info as to the current platform (platform lineage?), and a list of the supported platforms that the profile supports.

EDIT: ahh, I see you do include the backend platform.  Mentioning the profile platform might be nice, though.  Not a blocker.

clintoncwolfe commented on lib/inspec/profile.rb:
```diff
@@ -162,7 +168,8 @@ def params
 
     def collect_tests(include_list = @controls)
-      if !@tests_collected
+      unless @tests_collected
```
Hrm.  How would we get here, if we're on an unsupported platform?  I thought the code in runner.rb would catch the unsupported platform, issue a warning, and skip adding the profile to the test collector.  Is there another way to get here?

jquick replied:
I think we need this as the runner only passes in the parent but this will call itself for each dependent.

clintoncwolfe reviewed (commented):
No major blockers, just some clarifications and one desired extra test.  Looks great overall - great contribution!

miah committed:
Resolve tests using a supported platform different from testing platform

Add a control to `test/unit/mock/profiles/complete-profile` that would
work on any OS with a Internet connection. This allows the profile
to execute on any OS with success. `filesystem_spec.rb` was a control
that would only work on Linux and some BSD's.

We want profile tests to consistently work across development and testing
platforms, and not get 'skipped' in some cases.  Travis-CI tests on Linux,
Inspec Dev team uses Linux and MacOS, Appveyor tests on Windows

Also Updated `file_provider_test.rb` for `complete-profile` content changes.

If you `MockLoader.load_profile` on a unsupported platform you might not
hit the usual skip. Lets handle situations where the tests array in
Profile#load_checks_params could be nil.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Use safe navigation rather than checking if tests is nil.
Update tests to point to unsupported_inspec and account for WARN changes.
Make unsupported_inspec profile support os-family 'unsupported_inspec'

Signed-off-by: Miah Johnson <miah@chia-pet.org>

jerryaldrichiii reviewed (approved):
Looks good to me!

jquick reviewed (approved):
Thanks @miah
